### DISCUSSION TITLE
Do not update to `@main` by default

### DISCUSSION THREAD

**Discussion by laurentsimon** [renovatebot/renovate#16768] [Suggest an Idea]
Pinning a GHA by hash an/action@<hash> is the recommended way on Github, see https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions.

All the examples online, including in the https://github.com/actions/starter-workflows/ repo, uses this. Nobody adds a comment like `# tag=v1.2.3`. Yet, by default, renovatebot treats the lack of tag comment as meaning "@main". I think this behavior is unfortunate:

1. It requires work for developers to migrate.
2. The vast majority of uses case (99.999%) is intended to use a stable release, not at head
3. Main branch is a development branch, which means code may be *unstable*. Renovatebot should not update users to main branch unless they ask for it *explicitly*.
4. Services that help users to migrate to using hash, like https://app.stepsecurity.io/ (which is the recommended option by OSSF scorecard), also don't add this tag. 

By default, the lack of tags should mean that users want to get updates for releases. That should be the default behavior.
Users who want to pull updates from main should "opt-in" for this rare use case.

/cc @varunsh-coder

**Comment by rarkins** [comment]
I don't agree with this suggestion. The fact that Renovate has devised an elegant way to provide visibility into tags while also pinning to digests should be embraced. I think you should instead be lobbying to standardize such behavior in similar tools. If step security were to pin the tag in a comment like we suggest then you and others wouldn't have such a problem. A hash-only approach provides little visibility for the user into whether they are behind or not - they need to track down the tag which corresponds to their commit hash.

**Reply by laurentsimon** [reply]
I embrace the tags, I don't embrace that it defaults to `@main` by default without telling users. Most users don't realize it and they start pulling from the unstable HEAD. (This has happened for scorecard users; they started pulling from main which is unstable).

I think it would be fine for renovatebot to warn users that it cannot infer the version tag. I would be better than updating to `@main` by default.

/cc'ing @josepalafox for the standardization part.

**Reply by rarkins** [reply]
There's two separate scenarios we can consider:
- When Renovate does the digest pinning
- When some other tool has done the digest pinning

When I look at the repo you mentioned earlier, all these actions are already tied to a version tag: https://github.com/actions/starter-workflows/tree/main/.github/workflows

If Renovate pinned them to a digest, wouldn't it automatically include the tag in the comment?

So the problem is if you use a different tool to do that digest pinning which isn't as smart and just "loses" the tag and *replaces* it with only a hash.

My next question is, what is GitHub Actions behavior if someone depends on an action without a tag? e.g. `uses: pozil/auto-assign-issue`. In this case does it depend on `main`, or does actions seek out the most recent release or tag?

**Reply by laurentsimon** [reply]
> There's two separate scenarios we can consider:
> 
> * When Renovate does the digest pinning
> * When some other tool has done the digest pinning
> 
> When I look at the repo you mentioned earlier, all these actions are already tied to a version tag: https://github.com/actions/starter-workflows/tree/main/.github/workflows

Sorry for the confusion. The starter-workflow repository contains workflows that users can install via the GH "New Action" tab and "New scanning tool" tab. For example, if you visit `https://github.com/<your>/<repo>/actions/new`. All these workflows are automatically populated based on the template stored in this repo. If a user uses renovatebot and installs one of these workflows, renovatebot will assume the user wants to pull from `@main`.

> If Renovate pinned them to a digest, wouldn't it automatically include the tag in the comment?

It would. I'm more concerned with cases where the hash is already present.

> So the problem is if you use a different tool to do that digest pinning which isn't as smart and just "loses" the tag and _replaces_ it with only a hash.

correct.

> My next question is, what is GitHub Actions behavior if someone depends on an action without a tag? e.g. `uses: pozil/auto-assign-issue`. In this case does it depend on `main`, or does actions seek out the most recent release or tag?

This is considered an invalid workflow and it won't run: `Missing version for action` and `the 'uses' attribute must be a path, a Docker image, or owner/repo@ref`

**Reply by rarkins** [reply]
So when the other tools take a tagged action and convert it to a hash without annotating the tag with a version, they're essentially "discarding" the tag during that conversion. I'd really urge them not to do that. 

When we see a non annotated hash it might not even refer to a tag at all, or it might refer to multiple tags. 

Additionally "just taking the latest tag instead of main" isn't foolproof either. What if that tag was "unstable" or "3.0.0-alpha.1"? I know it won't be worse than today, but it's still not guaranteeing the stability you're hoping for. Far better is to use the combination of stable tag annotation plus hash which we recommend.

**Reply by laurentsimon** [reply]
> So when the other tools take a tagged action and convert it to a hash without annotating the tag with a version, they're essentially "discarding" the tag during that conversion. I'd really urge them not to do that.

it would be nice yes. Oftentimes, there is no tag or comment present. In cases where someone already uses hashes and then install renovatebot (or has renovatebot and uses a starterworkflow), it's too late. I think a conservative approach where renovatebot warns the user and tells them they must provide a `# tag=` may be a viable option. A users who wants to pull from main needs to make that explicit using `@main`: in this case it renovatebot need not add the hashes, I suppose 

> When we see a non annotated hash it might not even refer to a tag at all, or it might refer to multiple tags.
> 
> Additionally "just taking the latest tag instead of main" isn't foolproof either. What if that tag was "unstable" or "3.0.0-alpha.1"? I know it won't be worse than today, but it's still not guaranteeing the stability you're hoping for. Far better is to use the combination of stable tag annotation plus hash which we recommend.

I know it's a difficult problem to automate, which is why I feel warning users and asking them explicitly what they want may be a viable option to consider.

**Reply by rarkins** [reply]
I think we can log a warning whenever actions are updated from default branch (e.g. main) when tags are non present.

BTW GitHub could also help fix this problem by formalizing either the tag comment or better yet allowing both tag and hash to be specified in a `uses`, like Docker tags+digests can be

**Reply by laurentsimon** [reply]
Yes, docker style would be great. I've filed a feature request.

**Reply by laurentsimon** [reply]
> I think we can log a warning whenever actions are updated from default branch (e.g. main) when tags are non present.

In the PR description?

**Reply by rarkins** [reply]
We don't yet show warnings in PRs - they show up in the dependency dashboard

**Reply by laurentsimon** [reply]
ah, I was not aware of the dashboard. In general, I think it's good to provide information / context "inline", ie in the PR description since that's what developers look at during the course of merging a PR.

**Reply by rarkins** [reply]
Created #16924
